### ISSUE TITLE
Enhancement: Real-time acceleration updates via Debezium CDC 

### ISSUE THREAD

**Issue by phillipleblanc** [spiceai/spiceai#1785]
## Goal-State/What/Result

A new Debezium DataConnector that requires a non-Arrow acceleration engine with `refresh_mode: changes` will subscribe to changes in real-time (including the original snapshot) from a Kafka topic and apply those changes to the accelerated table to keep it updated in real-time with the original table.

## Why/Purpose

This is the holy grail of functionality we want to enable - a no compromises solution that combines both high-performance accelerated queries with highly efficient real-time delta updates.

## By When

**Issue/Spec written and reviewed:** 06/24
**Done-Done:** By `v0.15-alpha`, 06/27

## Done-Done

- [x] [First Principles Driven](https://github.com/spiceai/spiceai/blob/trunk/docs/PRINCIPLES.md)
- [x] The Algorithm
- [x] PM/Design Review
- [x] UX Review
- [x] Threat Model / Security Review
- [x] Tests
- [x] Documentation
- [x] Samples/Quickstarts/Tutorials
- [x] Release Notes

## The Algorithm

- [x] Every requirement questioned?
- [x] Delete (Scope) any part you can.
- [x] Simplify.
- [x] Break down into smaller iterations/milestones.
- [x] Opportunities for automation.

## Spec

```yaml
datasets:
- from: debezium:my_debezium_kakfa_topic
  name: cdc_table
  params:
    debezium_transport: kafka
    debezium_message_format: json # avro in the future
    # Most kafka consumers use `bootstrap.servers` as the parameter name 
    # https://docs.confluent.io/platform/current/installation/configuration/consumer-configs.html#bootstrap-servers
    kafka_bootstrap_servers: broker1:9092,broker2:9092,broker3:9092
  acceleration:
    enabled: true # acceleration is required for the debezium connector, it is an error to not specify an acceleration
    engine: duckdb # arrow isn't supported
    refresh_mode: changes # optional, default is `changes` - this is required to be set to `changes`, any other value is an error.
    mode: file # persistence is recommended to not have to rebuild the table each time Spice starts
    primary_key: id # We can discover the primary key of the table from the Debezium messages - but it can be optionally specified here to override the default.
    on_conflict:
      id: upsert # this on_conflict section is always set to `upsert` for the `primary_key` - so it is optionally to set here. If a user manually specifies something else - that is an error.
```

In the initial scope, we will only support JSON payloads from Debezium. Debezium also supports encoding the payloads as Arvo, which is much more efficient.

Kafka uses [consumer groups] to keep track of which messages in the stream a particular consumer has already consumed. Spice will auto-generate a unique consumer group id to ensure it is the only consumer of the changes stream. When combined with file-based persistence for acceleration tables, this could lead to a situation where Spice is restarted, Spice generates a new consumer group id and starts replaying changes against an accelerated table which already has those changes. To avoid this, Spice will create a new table in the accelerator engine called `spice_sys_metadata` which will store the consumer group id. Upon restart, Spice will check if a consumer group id already exists and use it first before generating a unique one.

[consumer groups]: https://docs.confluent.io/kafka/design/consumer-design.html#consumer-groups-group-ids-and-coordinators

Debezium supports schema evolution - that is out of scope for this initial feature.

## How/MAP

We will require [`at-least-once`](https://github.com/fede1024/rust-rdkafka/blob/master/examples/at_least_once.rs) delivery semantics to ensure no changes are missed. In general it is fine to receive a single message more than once, since we are treating write operations as `upserts` and deleting a non-existing row is fine. As long as we receive all of the messages in order, there will be no data quality issues.

On startup, we can be in two situations:
1) This is a newly instantiated dataset with no prior existence. A new consumer group will be generated and Spice will start reading messages from the stream from the beginning. We need to inspect the first message to build the Arrow schema that represents the table and find if there are any primary keys.
If no primary_key can be detected and none was specified in the spicepod, then this is an error.
If there are no messages (this is an empty topic), then Spice will keep waiting on the dataset initialization until a message becomes available)
We store the consumer group and primary key(s) into `spice_sys_metadata`
2) This is an already existing dataset with file persistence. We will read the existing consumer group from the `spice_sys_metadata` table and start processing messages from our last read message. We will inspect the schema from the accelerated table and the primary_key from the `spice_sys_metadata` table.

### Implementation (issues)

- [x] Spicepod update to add `RefreshMode::Changes`
- [x] Add Debezium DataConnector and require an acceleration.
- [x] Build Arrow schema from Debezium schema in the ChangeEvent message.
- [x] Store consumer group id and primary_key into `spice_sys_metadata` and load on startup.
- [x] Schema inference from existing accelerator table on startup.
- [x] Validate spicepod conforms to expected format.
- [x] Warn when using an acceleration engine without persistence.
- [x] Require `at-least-once` semantics in the code.

### QA

Integration test and manual testing by the team. Also during the next bug bash.
